Better Changelogs

Destructor-Ben opened the pull request:
### What is the new feature?
Slightly improved the automatic changelogs text, added substitution for the mod version, mod homepage, tML version, and tML build purpose into the changelogs, and made the automatic changelogs text optional - the automatic changelogs will only be created if there isn't anything in `changelog.txt`, this way people that prefer the automatic changelogs but still want to have customisable changelogs can still use them by pasting the default prompt in.

### Why should this be part of tModLoader?
Currently, changelogs are annoying if you don't like the default one because you have to delete it yourself and if you want to include the version, homepage, and other info it has to be manually copied over.

### Are there alternative designs?
Making the automatic changelogs disabled with a property in build.txt/csproj.

### Sample usage for the new feature
In `changelog.txt`:
```
Example Mod v{ModVersion} for tModLoader v{tMLVersion}
- Example change 1
- Example change 2
```

### ExampleMod updates
<!-- If you also updated ExampleMod for your new feature, let us know here -->
Added an example changelog that I believe doesn't get used when it is published.
Updated the changelog hardcoded in `build.yml`, but I'm not sure if the substitution works there since I don't know the automated uploading works.

Destructor-Ben committed:
Update WorkshopHelper.cs.patch

Solxanich commented on patches/tModLoader/Terraria/Social/Steam/WorkshopHelper.cs.patch:
```diff
@@ -289,13 +289,17 @@
 +
 +					string refs = _entryData.BuildData["workshopdeps"];
-+
++					
```
Weird  line change?

Solxanich commented on patches/tModLoader/Terraria/Social/Steam/WorkshopHelper.cs.patch:
```diff
@@ -289,13 +289,17 @@
-+						patchNotes += $" learn more @ { _entryData.BuildData["homepage"]}";
-+					patchNotes += $"\n{_entryData.ChangeNotes}";
++					patchNotes = _entryData.ChangeNotes;
++					if (string.IsNullOrEmpty(patchNotes)) {
```
Please add a comment indicating we will supply our own patchNotes if the modder doesn't have anything in the changelog.txt

Solxanich commented on patches/tModLoader/Terraria/Social/Steam/WorkshopHelper.cs.patch:
```diff
@@ -289,13 +289,17 @@
++							patchNotes += ", learn more at {1}";
++					}
++					// {0} to mod version, {1} to mod homepage, {2} to tModLoader version, {3} to tModLoader build purpose
++					patchNotes = string.Format(patchNotes, _entryData.BuildData["trueversion"], _entryData.BuildData["homepage"], BuildInfo.tMLVersion.MajorMinor(), BuildInfo.Purpose);
```
Per your request, could consider the LocalizedText.FormatWith solution.

Something like, although probably more elegant code:
``for int i = 0 to 4 {
patchNotes = patchNotes.Replace($"{identifier[i]}, value[i]);
}``
And use arrays. 

Because, yes, it is very confusing to know what the numbers mean.

Solxanich commented on patches/tModLoader/Terraria/Social/Steam/WorkshopHelper.cs.patch:
```diff
@@ -289,13 +289,17 @@
++						if (!string.IsNullOrEmpty(_entryData.BuildData["homepage"]))
++							patchNotes += ", learn more at {1}";
++					}
++					// {0} to mod version, {1} to mod homepage, {2} to tModLoader version, {3} to tModLoader build purpose
```
Please update example mod changelog.txt to match the new format as part of PR.

Solxanich reviewed (commented):
I would like to see the ExampleMod ChangeLog.txt update to see how confusing the numbered replace format looks for a new modder. 
I believe it is hardcoded in the build.yaml? Include an example changelog.txt in its folder regardless if not present.

Destructor-Ben committed:
review

Destructor-Ben committed:
Fix EM changelog

not sure it works bc i can't really test it and I have no idea what the build.yml ciprep variable is

Destructor-Ben committed:
Add example changelog.txt

JavidPack commented:
Is this feature working?

https://steamcommunity.com/sharedfiles/filedetails/changelog/2615761271

![image](https://github.com/tModLoader/tModLoader/assets/4522492/6edbad17-4dc7-4aca-b75c-6d838736bb87)

Destructor-Ben commented:
It definitely works for normal mods (can sort mod browser to newest and check a changelog), but I'm pretty's sure ExampleMod manually populates the change log when being published because (like you said) it's published weirdly.
